Improvements to the StatusView

jgmdev opened the pull request:
Features implemented:

* Support for predicates by introducing add_item().
* Support for performing click actions on items.
* Support for optional tooltips on item hover.
* Deprecate the usage of get_item().

~~Other stuff for the future that I'm not sure could implement now~~:

- [x] Hide items that are too huge by rendering with `clip_rect`
- [x] Ability to drag the left or right if too many items to display.
- [x] Properly detect the visible part of an item for click and mouse hover events.

## Demonstration

https://user-images.githubusercontent.com/1702572/154910282-2fb91dc3-d898-40fb-b560-be371f2be7c1.mp4

jgmdev commented:
Forgot to mention that I also documented parts of the StatusView code using EmmyLua which can serve as an experiment of how one would document the core code directly.

jgmdev commented:
Restructured almost everything that I did before, borrowed some ideas from vsc statusbar api and Implemented some more features as described on latest commit, for convenience will also paste them here:

* Items are now objects that can be retrieved and manipulated.
* clip rect is used for left and right panes
* initial support for items to do their own custom drawing (this was takase idea) by also
  doing a clip rect for them
* a custom background color can be specified for the item.
* a command or function can be executed on item click.
* Introduced functionality to easily hide or show all or specific items.
* Better handling of deprecated get_items()
* Spacing is automatically added to items and cleaned on deprecated
  items.
* Default items where separated and given the names:
  doc:file, doc:position, doc:indentation, doc:lines,
  doc:line-ending, core.commandview.
* Some default right or left click actions where given to the default
  items.
* Started adding required bits to support dragging to left and right
  panes when some items aren't visible.

Note: should work well enough already but maybe some repetitive stuff can be
cleaned out.

Also here a screen capture showing some of the things that can be done now:


https://user-images.githubusercontent.com/1702572/155292956-2400e2f7-aeb2-4e06-b7bc-64d001ec0972.mp4

jgmdev commented:
Dragging and scrolling left or right panels is now supported, now last missing bit is detecting the visible area of an item for click and mouse hover events:


https://user-images.githubusercontent.com/1702572/155428960-fece7ffa-88f7-49c5-b47e-c87f4518c36a.mp4

jgmdev commented:
For reference, leaving some sample code on how to do custom drawing into the status bar:

```lua
if not core.status_view:get_item("my:custom-item") then
  local item = core.status_view:add_item(nil, "my:custom-item", StatusView.Item.LEFT)
  item.on_draw = function(x, y, h, calc_only)
    --true when the status bar needs to know the width of drawing area
    if calc_only then
      return 100
    end
    --if calc_only not true is time to draw
    renderer.draw_text(style.font, "Hello", x, y, style.text)
  end
  item.background_color = { common.color "#0f6773" }
  item.tooltip = "my item"
end
```

Guldoman reviewed (commented):
What do you think about having an hovered background color for items with `on_click` commands?
If so, an `Item.background_color_hovered` would be needed. Also the hovered status should be passed to `on_draw`.

This isn't too important, but resizing doesn't gracefully update the scrolled position:

https://user-images.githubusercontent.com/2798487/155843090-a3f7d577-9a8c-4d61-b012-ea20b5922133.mp4

jgmdev commented:
> What do you think about having an hovered background color for items with `on_click` commands? If so, an `Item.background_color_hovered` would be needed. Also the hovered status should be passed to `on_draw`.

Implemented:

https://user-images.githubusercontent.com/1702572/156065762-7d88de2a-8382-486d-852b-525ea95dca28.mp4

> This isn't too important, but resizing doesn't gracefully update the scrolled position:

Fixed (sorry for flickering on preview seems my mesa opengl driver has wayland issues with sdl when doing that):

https://user-images.githubusercontent.com/1702572/156065721-8722415e-9b51-47a9-856d-e0e491ca5300.mp4

jgmdev commented on data/core/statusview.lua:
```diff
@@ -104,60 +466,531 @@ function StatusView:draw_items(items, right_align, yoffset)
+---@return table right
+function StatusView:get_items(nowarn)
+  if not nowarn and not self.get_items_warn then
+    core.error(
```
yep, and change the error icon maybe to `X` and use the error icon `!` for warnings with some orange or yellow color.
